src/mono/wasm/runtime/invoke-js.ts

```diff
@@ -0,0 +1,316 @@
+        //console.log("mono_wasm_invoke_js_1 C");
+        const res_js = fn(arg1_js);
+        //console.log("mono_wasm_invoke_js_1 D");
+        covert_result_to_cs(signature, data, res_js);
+        //console.log("mono_wasm_invoke_js_1 E");
+        return InvokeJSResult.Success;
+    } catch (ex) {
+        //console.log("mono_wasm_invoke_js_1 F");
+        _marshal_exception_to_cs(data, ex);
+        return InvokeJSResult.Exception;
+    }
+}
+
+export function mono_wasm_bind_js_function(interned_function_id: MonoString, arg_count: number, signature: VoidPtr, function_js_handle: Int32Ptr, is_exception: Int32Ptr): MonoObject {
+    try {
+        const fn = mono_wasm_lookup_function(interned_function_id);
+        const closure: any = { MONO, BINDING, INTERNAL, fn };
+        let body = "try {\n";
+
+        for (let position = 1; position <= arg_count; position++) {
+            const offset = position << 2;
+            const mono_type = <MonoType><any>getU32(<any>signature + offset);
+            const covertor = cs_to_js_map.get(mono_type);
+            assert(covertor, `Unknow covertor for type ${mono_type} at ${position}`);
+            const convertor_name = `covertor${position}`;
+            closure[convertor_name] = covertor;
+
+            body += `  const arg${position}_js = closure.${convertor_name}(<any>data + ${offset});\n`;
+        }
+        if (arg_count === 0) {
+            body += "  const res_js = closure.fn();\n";
+        }
+        else if (arg_count === 1) {
+            body += "  const res_js = closure.fn(arg1_js);\n";
+        }
+
+        const res_mono_type = <MonoType><any>getU32(<any>signature);
+        if (res_mono_type === MonoTypeNull) {
+            body += "  assert(value === undefined, 'Function returned unexpected value, C# signature is void');\n";
+        } else {
+            const covertor = js_to_cs_map.get(res_mono_type);
+            assert(covertor, `Unknow covertor for type ${res_mono_type} at ret`);
+            closure["covertor0"] = covertor;
+
+            body += "  closure.covertor0(<any>data);\n";
+        }
+
+        body += "  return InvokeJSResult.Success;\n";
+        body += "} catch (ex) {\n";
+        body += "  _marshal_exception_to_cs(data, ex);\n";
+        body += "  return InvokeJSResult.FunctionThrewException;\n";
+        body += "}\n";
+
+        const bound_fn = new Function("closure", "data", body).bind(null, closure);
+        const bound_function_handle = mono_wasm_get_js_handle(bound_fn)!;
+        setI32(function_js_handle, <any>bound_function_handle);
+        return MonoObjectNull;
+    } catch (ex) {
+        return wrap_error(is_exception, ex);
+    }
+}
+
+export function mono_wasm_invoke_bound_function(function_js_handle: JSHandle, data: VoidPtr): InvokeJSResult {
+    const js_fn = mono_wasm_get_jsobj_from_js_handle(function_js_handle);
+    assert(js_fn, `Unknow function for handle ${function_js_handle}`);
+    return js_fn(data);
```

Comment:
What this function actually does is a little unclear from the names. If I understand it correctly, what's happening is it pulls a bound function's *wrapper* out of the handle table, and invokes the wrapper which is responsible for actually invoking the js function. so "js_fn" is a misleading name for this and I struggled at first to understand why we would be passing a single VoidPtr to an arbitrary javascript function.

My proposed names would be 'wrapper_js_handle' for the handle (so it's clear that it's not a handle to the *actual* js function target) and 'wrapper' for the local. We may also need to do a check here of some sort to ensure that the jsobj we got is actually a bound function and not something else, since they appear to live in a global handle namespace and someone writing broken code could accidentally pass the wrong kind of handle to the invoke icall. (We could just set a property on the wrapper function objects to check for)

Reply:
I added TODO for the check and renamed the handle to `bound_function_js_handle`